### DISCUSSION TITLE
Uniaxial tensile simulation of polycrystalline model based on crystal plasticity

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#17423] [Q&A Modules: Solid mechanics] [answered]
Dear MOOSE experts,
I recently wrote an input file based on the files [prop_grain_read.i](https://github.com/PengWei97/monkey/blob/main/problem/poly_cp/prop_grain_read.i) and [crysp_user_object.i](https://github.com/PengWei97/monkey/blob/main/problem/poly_cp/crysp_user_object.i) to simulate plastic uniaxial stretching of polycrystalline crystals, and established a two-dimensional model of 0.21*0.21mm. There are 200 grains in it. Stretching in the y-axis direction. 
284 time steps, time = 0.327438, stretching by 0.03%.
But the calculation is very slow. Is there any suggestion to speed up the calculation?

The content of the written input file and output terminal is as follows,
[prop_grain_read_cp.i](https://github.com/PengWei97/monkey/blob/main/problem/poly_cp/prop_grain_read_cp.i)
[prop_grain_read_cp.txt](https://github.com/idaholab/moose/files/6198217/prop_grain_read_cp.txt)

Any suggestions or recommendations to fix these problems would be greatly appreciated.
Thank you
Wei Peng

**Comment by jiangwen84** [comment, marked as answer]
@PengWei97 I need to let you know that we are currently switching the UO-based CP (the one you used) to stress-update based CP. One on-going PR is https://github.com/idaholab/moose/pull/17405. The new version might give you a bit speedup. 

For your current input file, I suggest you add `maximum_substep_iteration = 4` in your `FiniteStrainUObasedCP` block. It will allow you to take larger `dt`. With sub stepping, each time step will take longer time, but it can speed up your simulation by taking less number of time steps. 

@sapitts @dewenyushu  any other thoughts?

**Reply by dewenyushu** [reply]
Hi @PengWei97 I agree with @jiangwen84 that using sub-stepping should allow larger `dt` thus speed up computation in general. 

For the new stress-update based CP base class  (#17405). As Wen mentioned, this is a new implementation which will be more actively developed and maintained than the UO-based version in the future. With this new class, we do see some improvements in terms of the total linear/nonlinear iterations for several existing models (and some of them are improved a lot!). However, we are yet to quantify the improvement or examine the consistency of this improvement among all existing models.

**Reply by PengWei97** [reply]
Hi @jiangwen84 ,Thank you for your guidance.

I added `maximum_substep_iteration = 5` to the` FiniteStrainUObasedCP `block according to your suggestion. But it is still calculated at the minimum incremental step set in the seventh step, and it took five and a half hours to calculate to `"Time Step 805, time = 0.897438, dt = 0.001"`. Similarly, I used umat calculations on abaqus, and it took less than half an hour to complete the uniaxial stretching simulation. I think it must be an unreasonable setting in that place. The following is the input file. Do you have any suggestions for improvement?

`[Mesh]
  type = GeneratedMesh
  dim = 2
  elem_type = QUAD4
  displacements = 'disp_x disp_y'
  nx = 100
  ny = 100
  xmax = 0.21
  ymax = 0.21
[]

[Variables]
  [./disp_x]
    block = 0
  [../]
  [./disp_y]
    block = 0
  [../]
[]

[GlobalParams]
  volumetric_locking_correction=true
[]

[AuxVariables]
  [./stress_yy]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
  [./e_yy]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
[]

[Functions]
  [./tdisp]
    type = ParsedFunction
    value = 2e-4*t # Approximately 0.1% loading per second
  [../]
[]

[UserObjects]
  [./prop_read]
    type = ElementPropertyReadFile
    # need read
    prop_file_name = 'input_file.txt' # euler angle
    # Enter file data as prop#1, prop#2, .., prop#nprop
    nprop = 3
    read_type = grain
    ngrain = 200
    # rand_seed = 25346
    rve_type = periodic
    # Periodic or non-periodic grain distribution: Default is non-periodic
  [../]
[]

[AuxKernels]
  [./stress_yy]
    type = RankTwoAux
    variable = stress_yy
    rank_two_tensor = stress
    index_j = 1
    index_i = 1
    execute_on = 'initial timestep_end'
    block = 0
  [../]
  [./e_yy]
    type = RankTwoAux
    variable = e_yy
    rank_two_tensor = lage
    index_j = 1
    index_i = 1
    execute_on = timestep_end
    block = 0
  [../]
[]

[BCs]
  [./fix_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'left'
    value = 0
  [../]
  [./fix_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom'
    value = 0
  [../]
  [./tdisp]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function = tdisp
  [../]
[]

[UserObjects]
  [./slip_rate_gss]
    type = CrystalPlasticitySlipRateGSS
    variable_size = 12
    slip_sys_file_name = input_slip_sys.txt # slip system
    num_slip_sys_flowrate_props = 2
    flowprops = '1 4 0.001 0.1 5 8 0.001 0.1 9 12 0.001 0.1'
    uo_state_var_name = state_var_gss
  [../]
  [./slip_resistance_gss]
    type = CrystalPlasticitySlipResistanceGSS
    variable_size = 12
    uo_state_var_name = state_var_gss
  [../]
  [./state_var_gss]
    type = CrystalPlasticityStateVariable
    variable_size = 12
    groups = '0 4 8 12'
    group_values = '60.8 60.8 60.8'
    uo_state_var_evol_rate_comp_name = state_var_evol_rate_comp_gss
    scale_factor = 1.0
  [../]
  [./state_var_evol_rate_comp_gss]
    type = CrystalPlasticityStateVarRateComponentGSS
    variable_size = 12
    hprops = '1.0 541.5 109.8 2.5'
    uo_slip_rate_name = slip_rate_gss
    uo_state_var_name = state_var_gss
  [../]
[]

[Materials]
  [./crysp]
    type = FiniteStrainUObasedCP
    stol = 1e-2
    tan_mod_type = exact
    uo_slip_rates = 'slip_rate_gss'
    uo_slip_resistances = 'slip_resistance_gss'
    uo_state_vars = 'state_var_gss'
    uo_state_var_evol_rate_comps = 'state_var_evol_rate_comp_gss'
    maximum_substep_iteration = 5 # add for an advice of jiangwen84
  [../]
  [./strain]
    type = ComputeFiniteStrain
    displacements = 'disp_x disp_y'
  [../]
  [./elasticity_tensor]
    type = ComputeElasticityTensorCP
    C_ijkl = '1.684e5 1.214e5 1.214e5 1.684e5 1.214e5 1.684e5 0.754e5 0.754e5 0.754e5'
    fill_method = symmetric9
    read_prop_user_object = prop_read
  [../]
[]

[Postprocessors]
  [./stress_yy]
    type = ElementAverageValue
    variable = stress_yy
    block = 'ANY_BLOCK_ID 0'
  [../]
  [./e_yy]
    type = ElementAverageValue
    variable = e_yy
    block = 'ANY_BLOCK_ID 0'
  [../]
[]

[Preconditioning]
  [./smp]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient

  #Preconditioned JFNK (default)
  solve_type = 'PJFNK'

  l_max_its = 30
  l_tol = 1.0e-4
  nl_max_its = 20
  nl_rel_tol = 1.0e-9

  end_time = 1
  dtmin = 0.001

  [./TimeStepper]
    type = IterationAdaptiveDT
    dt = 0.05 
    optimal_iterations = 6 
  [../]
[]

[Outputs]
  file_base = prop_grain_read_out_xiu02
  exodus = true
  csv = true
[]

[Kernels]
  [./TensorMechanics]
    displacements = 'disp_x disp_y'
    use_displaced_mesh = true
  [../]
[]
`

**Reply by PengWei97** [reply]
> Hi @PengWei97 I agree with @jiangwen84 that using sub-stepping should allow larger `dt` thus speed up computation in general.
> 
> For the new stress-update based CP base class (#17405). As Wen mentioned, this is a new implementation which will be more actively developed and maintained than the UO-based version in the future. With this new class, we do see some improvements in terms of the total linear/nonlinear iterations for several existing models (and some of them are improved a lot!). However, we are yet to quantify the improvement or examine the consistency of this improvement among all existing models.

Hi @dewenyushu,Thank you for your advice.
I tried to clone the relevant *.C and *.h files in my development branch into the application I created, but an error was displayed when compiling, the error is as follows,
[error_make.txt](https://github.com/idaholab/moose/files/6210004/error_make.txt)

Is it because this class is still under development, so we can't use it now, or is it for other reasons?

**Reply by jiangwen84** [reply]
Try one more thing: 

add 

  petsc_options_iname = '-pc_type' 
  petsc_options_value = 'lu'

to `[Executioner]` block.

**Reply by PengWei97** [reply]
> Try one more thing:
> 
> add
> 
> petsc_options_iname = '-pc_type'
> petsc_options_value = 'lu'
> 
> to `[Executioner]` block.

hi @jiangwen84 ，follow your suggestion, this problem has been solved.
Thank you very much for your guidance.
wei

**Reply by dewenyushu** [reply]
> > Hi @PengWei97 I agree with @jiangwen84 that using sub-stepping should allow larger `dt` thus speed up computation in general.
> > For the new stress-update based CP base class (#17405). As Wen mentioned, this is a new implementation which will be more actively developed and maintained than the UO-based version in the future. With this new class, we do see some improvements in terms of the total linear/nonlinear iterations for several existing models (and some of them are improved a lot!). However, we are yet to quantify the improvement or examine the consistency of this improvement among all existing models.
> 
> Hi @dewenyushu,Thank you for your advice.
> I tried to clone the relevant *.C and *.h files in my development branch into the application I created, but an error was displayed when compiling, the error is as follows,
> [error_make.txt](https://github.com/idaholab/moose/files/6210004/error_make.txt)
> 
> Is it because this class is still under development, so we can't use it now, or is it for other reasons?

Thanks for trying this out! By looking at your log file, it seems like you will need to checkout the Kalidindi class (`CrystalPlasticityKalidindiUpdate.C` and .h) from my PR as well since I have made some changes for it being able to run with the new base class. 

The new base class is something that is still in development right now and transforming all existing CP models to the new base class will require some more work. Currently, only the Kalidindi CP model works with this new base class. So I will not be surprised if there being some glitches when transforming your model without some changes. If you come across some new issues, please keep us updated.

**Reply by PengWei97** [reply]
> > > Hi @PengWei97 I agree with @jiangwen84 that using sub-stepping should allow larger `dt` thus speed up computation in general.
> > > For the new stress-update based CP base class (#17405). As Wen mentioned, this is a new implementation which will be more actively developed and maintained than the UO-based version in the future. With this new class, we do see some improvements in terms of the total linear/nonlinear iterations for several existing models (and some of them are improved a lot!). However, we are yet to quantify the improvement or examine the consistency of this improvement among all existing models.
> > 
> > 
> > Hi @dewenyushu,Thank you for your advice.
> > I tried to clone the relevant *.C and *.h files in my development branch into the application I created, but an error was displayed when compiling, the error is as follows,
> > [error_make.txt](https://github.com/idaholab/moose/files/6210004/error_make.txt)
> > Is it because this class is still under development, so we can't use it now, or is it for other reasons?
> 
> Thanks for trying this out! By looking at your log file, it seems like you will need to checkout the Kalidindi class (`CrystalPlasticityKalidindiUpdate.C` and .h) from my PR as well since I have made some changes for it being able to run with the new base class.
> 
> The new base class is something that is still in development right now and transforming all existing CP models to the new base class will require some more work. Currently, only the Kalidindi CP model works with this new base class. So I will not be surprised if there being some glitches when transforming your model without some changes. If you come across some new issues, please keep us updated.

Thank you very much. I will try to modify it.
